### DISCUSSION TITLE
Bridging the gap between a `XxxResult` class and `TypedResult`

### DISCUSSION THREAD

**Discussion by svarlet** [simolus3/drift#1451] [Q&A] [answered]
Hi,

I've made this method:
```dart
DownloadProxy _readDownloadProxy(TypedResult row) {
    final download = row.readTable(downloads);
    final syncpoint = row.readTable(syncpoints);
    // ...
    return DownloadProxy(download, location, this);
  }
```

I just wrote a select query inside my `.moor` file:
```sql
-- content omitted for clarity
getResumables:
    SELECT * FROM downloads AS d ... WHERE ...; 
```

When I type the following:
```dart
final rows = await getResumables().get();
for (final row in rows) {
    final resumable = _readDownloadProxy(row);
    // ...
}
```
the compiler complains that `GetResumableResult` does not match the type `TypedResult`. So I can't reuse my convenient `_readDownloadProxy(...)` method, and I can't call `readTable(row)` either since `GetResumableResult` does not inherit `TypedResult`. What would you recommend doing?

**Comment by svarlet** [comment]
I'm interested to understand how the ideal solution will scale (in terms of code duplication) considering that a `XxxResult` class does not extend or implement anything. I guess that N Result classes would require N almost identical mappers/converters?

**Comment by simolus3** [comment, marked as answer]
We do have the `raw_result_set_data` [builder option](https://moor.simonbinder.eu/docs/advanced-features/builder_options/) which can be used to obtain the underlying `QueryRow` from a result set.  However, it does not give you a `TypedResult`.

I think it will be hard to generate a `TypedResult` from a compiled statement in a moor file. A `TypedResult` is optimized towards usages coming from the Dart query builder and I think the additional flexibility of direct SQL maps poorly to that.
If you have lots of queries all returning the same data, you can customize the [name of the result class](https://moor.simonbinder.eu/docs/using-sql/moor_files/#result-class-names). If you give each query the same result class name, moor will only generate a single class for all of them.

Finally, you can use [nested results](https://moor.simonbinder.eu/docs/using-sql/moor_files/#nested-results) to avoid some of the pain of reading flat rows containing multiple tables. For instance, if you write

```sql
getResumables  AS Resumable: SELECT d.**, p.** 
  FROM downloads AS d 
    INNER JOIN syncpoints AS p ON ...
  WHERE ...; 
```

then moor will generate a class similar to this:

```dart
class Resumable {
  final Download d;
  final SyncPoint p;
}
```

(you can change the name of the generated fields by changing the alias in SQL)

**Reply by svarlet** [reply]
Thanks for the prompt answer. So your suggestion is to fully flesh out this common pair of values (a download and its associated syncpoint) with a dedicated class, and make queries return (list of) this type.
Do I understand correctly?

**Reply by simolus3** [reply]
You don't even have to write a dedicated class for this. If you write `AS Resumable`, moor is still going to write the `Resumable` class. It just allows you to use the same class for different queries with the same rows.
